### ISSUE TITLE
TypeError Exception when parsing sqlite file

### ISSUE THREAD

**Issue by romor** [uktrade/stream-sqlite#5]
I get the following exception when parsing a sqlite file:

```
  File "/home/roman/enline/cloud/lambda/sqldecoder.py", line 98, in decode
    for table_name, _pragma_table_info, rows in stream_sqlite(byte_stream, max_buffer_size):
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 441, in stream_sqlite
    for (name, info), single_table_rows in grouped_by_table:
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 438, in <genexpr>
    table_rows = ((table_name, table_info, row) for table_name, table_info, row in table_rows_including_internal if not table_name.startswith('sqlite_'))
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 425, in yield_table_rows
    yield from process_page(page_bytes, page_reader)
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 325, in process_table_page
    yield from \
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 304, in process_table_leaf_non_master
    yield from process_initial_payload(
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 172, in process_initial_payload
    yield from full_payload_processor(cell_num_reader, cell_varint_reader)
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 292, in process_non_master_leaf_row
    yield table_name, table_info, read_table_row(rowid, cell_num_reader, cell_varint_reader)
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 229, in read_table_row
    return row_constructor(rowid, *tuple(
  File "/usr/local/lib/python3.9/dist-packages/stream_sqlite.py", line 261, in <lambda>
    row_constructor = lambda rowid, *values: row_namedtuple(*tuple((rowid if i == rowid_alias_index else value) for i, value in enumerate(values)))
TypeError: <lambda>() missing 1 required positional argument: 'description'
```

This file was recently changed in structure by `ALTER TABLE alarming ADD COLUMN description TEXT`. Before this, the file was successfully parsed. So I assume that this change is the cause of above exception.

**Comment by michalc** [comment]
> This file was recently changed in structure by ALTER TABLE alarming ADD COLUMN description TEXT. Before this, the file was successfully parsed. So I assume that this change is the cause of above exception.

Ah that sounds likely. Looking at https://www.sqlite.org/fileformat.html I see 

> A record might have fewer values than the number of columns in the corresponding table. This can happen, for example, after an [ALTER TABLE ... ADD COLUMN](https://www.sqlite.org/lang_altertable.html) SQL statement has increased the number of columns in the table schema without modifying preexisting rows in the table. Missing values at the end of the record are filled in using the [default value](https://www.sqlite.org/lang_createtable.html#dfltval) for the corresponding columns defined in the table schema.

And don't think we handle this case properly. But -  suspect it's not a project to add support for this. We already extract metadata for tables, so it _should_ just be a few lines of code in the appropriate place, probably around https://github.com/uktrade/stream-sqlite/blob/967f25148c61e7df3d231c2f3ad6fc13f8b4da5f/stream_sqlite.py#L263

**Comment by romor** [comment]
Thanks a lot for the fast feedback!

> And don't think we handle this case properly. But - suspect it's not a project to add support for this.

Please let me know if you plan to support/fix this or not? As I use the library on a cloud server with different IoT devices sending such sqlite files, I need to solve this somehow. Maybe you could suggest a work-around?

Also, please let me know if I can contribute something. Of course, I could provide a sqlite file showing the problem. Also I might be able to help by contributing a pull request if you could advice and point me in a proper direction?

**Comment by michalc** [comment]
> Please let me know if you plan to support/fix this or not?

We do - and done! Released in [v0.0.41](https://pypi.org/project/stream-sqlite/0.0.41/)

> I use the library on a cloud server with different IoT devices sending such sqlite files

I'm curious, what sort of IoT devices? And if it's ok to ask, what's sort of data are in the sqlite files?

(Closing this issue since I suspect this is fixed, but feel free to re-open if this isn't the case)

**Comment by romor** [comment]
Awesome, thanks a lot! I can confirm that this fixes the issue.

We operate energy management devices that monitor and control electrical energy consumption (www.embyt.com). The sqlite databases store energy data and also alarming information, such as unexpected high consumption.
